### PR TITLE
Changes plutus builtin-type bytestring from lazy to strict. Fixes #1692

### LINKED ISSUE #1692: Ditch lazy bytestrings
Currently we have lazy `ByteString`s as constants, which really should be strict ones, because lazy ones get allocated in huge chunks and Plutus Core is a strict language, so no point in having lazy bytestring anyway.

### PR DESCRIPTION
**Pull request by bezirg**
Changes the plutus builtin type to be `Data.ByteString` aka strict bytestring, instead of `Data.ByteString.Lazy` .

This touches code among almost all projects. I've tried not to change any lazy bytestrings that do not correspond to actual builtin-types, e.g. input-ouput reading/writing files using lazy bytestrings. I may have touched more bytestrings than I should though (especially at scb @krisajenkins )

Tested for all projects, except metatheory (bcs I don't have agda installed). @jmchapman Can you take a look if metatheory still works?

I didn't touch any plutus code-examples in `plutus-book`. I guess that is fine because the examples are out-of-date anyway.

### CODE REVIEW

File: plutus-ledger/src/Ledger/Scripts.hs [hunk 1]
```diff
@@ -112,17 +112,17 @@ infrequently (I believe).
 -}
 instance Eq Script where
     {-# INLINABLE (==) #-}
-    a == b = serialise a == serialise b
+    a == b = BSL.toStrict (serialise a) == BSL.toStrict (serialise b)
```
Comment (michaelpj) [inline plutus-ledger/src/Ledger/Scripts.hs:115]: Lazy bytestrings have a pefectly good `Eq` instance, no?

File: plutus-tx/src/Language/PlutusTx/Data.hs [hunk 2]
```diff
@@ -61,10 +61,9 @@ viewMap (CBOR.TMap m)  = Just m
 viewMap (CBOR.TMapI m) = Just m
 viewMap _              = Nothing
 
-viewBytes :: CBOR.Term -> Maybe BSL.ByteString
-viewBytes (CBOR.TBytes b)  = Just (BSL.fromStrict b)
-viewBytes (CBOR.TBytesI b) = Just b
```
Comment (michaelpj) [inline plutus-tx/src/Language/PlutusTx/Data.hs:66]: This case is important!

**Review by michaelpj — APPROVED** [review]
This murders it throughout the codebase, not just in the builtin types, but I think that's fine.

@effectfully I remember you suggesting that we should maybe use `ShortByteString` for some of these. Good time to try it, perhaps?

**Comment by michaelpj** [comment]
Few more build failures, a couple of redundant imports that fail the CI with `-Werror`

**Comment by jmchapman** [comment]
I'm just working on a small fix to metatheory.

**Review by jmchapman — APPROVED** [review]
The `metatheory` changes look fine to me.

**Comment by effectfully** [comment]
> @effectfully I remember you suggesting that we should maybe use ShortByteString for some of these. Good time to try it, perhaps?

I think I mentioned it in the context of variable names representation (and I suggested `ShortByteString`-based `ShortText` (this probably was before I learned that regular `Text` uses unpinned memory) -- provided we can prove it actually improves memory footprint and is not too painful API-wise). I'll take a look at the PR, but I don't think we need that level of optimization anywhere and the [API](http://hackage.haskell.org/package/bytestring-0.10.12.0/docs/Data-ByteString-Short.html) of `ShortByteString` is just too tiny and everything else uses normal `ByteString`s, so we'd have to convert to regular `ByteString` anyway.

File: plutus-ledger/src/Ledger/Scripts.hs (hunk 1 above)
Comment (effectfully) [reply plutus-ledger/src/Ledger/Scripts.hs:115]: I'm also surprised.

**Review by effectfully — APPROVED** [review]
Huge upvote from me.

File: plutus-core/generators/Language/PlutusCore/Generators/Internal/Denotation.hs [hunk 3]
```diff
@@ -137,9 +137,9 @@ typedBuiltinNames
     . insertTypedStaticBuiltinName typedGreaterThanInteger   (>)
     . insertTypedStaticBuiltinName typedGreaterThanEqInteger (>=)
     . insertTypedStaticBuiltinName typedEqInteger            (==)
-    . insertTypedStaticBuiltinName typedConcatenate          (coerce BSL.append)
-    . insertTypedStaticBuiltinName typedTakeByteString       (coerce BSL.take . integerToInt64)
-    . insertTypedStaticBuiltinName typedDropByteString       (coerce BSL.drop . integerToInt64)
+    . insertTypedStaticBuiltinName typedConcatenate          (coerce BS.append)
```
Comment (effectfully) [inline plutus-core/generators/Language/PlutusCore/Generators/Internal/Denotation.hs:140]: While you're here, we don't need those `coerce`s anymore, do we?

File: plutus-core/src/Language/PlutusCore/Constant/Apply.hs [hunk 4]
```diff
@@ -205,13 +204,13 @@ applyStaticBuiltinName name args = do
         TakeByteString ->
             applyTypedStaticBuiltinName
                 typedTakeByteString
-                (coerce BSL.take . integerToInt64)
+                (coerce BS.take . integerToInt)
```
Comment (effectfully) [inline plutus-core/src/Language/PlutusCore/Constant/Apply.hs:207]: And in this file.

File: plutus-core/test/Evaluation/ApplyBuiltinName.hs [hunk 5]
```diff
@@ -137,7 +137,7 @@ test_typedConcatenate
 test_typedTakeByteString :: TestTree
 test_typedTakeByteString
     = testProperty "typedTakeByteString"
-    $ prop_applyStaticBuiltinName typedTakeByteString (coerce BSL.take . integerToInt64)
+    $ prop_applyStaticBuiltinName typedTakeByteString (coerce BS.take . integerToInt)
```
Comment (effectfully) [inline plutus-core/test/Evaluation/ApplyBuiltinName.hs:140]: And in this file.

File: plutus-core/untyped-plutus-core-test/Evaluation/ApplyBuiltinName.hs [hunk 6]
```diff
@@ -133,7 +133,7 @@ test_typedConcatenate
 test_typedTakeByteString :: TestTree
 test_typedTakeByteString
     = testProperty "typedTakeByteString"
-    $ prop_applyStaticBuiltinName typedTakeByteString (coerce BSL.take . integerToInt64)
+    $ prop_applyStaticBuiltinName typedTakeByteString (coerce BS.take . integerToInt)
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core-test/Evaluation/ApplyBuiltinName.hs:136]: And in this file.

**Review by effectfully — COMMENTED** [review]
I think we have some redundant stuff left from some previous version of builtins.

**Comment by michaelpj** [comment]
Some issue with the PS generation, summon @krisajenkins ?

Looks like hashes change, so you need to update the test output for `plutus-use-cases` which has hashes in. `cabal test plutus-use-cases --test-options=--accept` will do it.

**Review by michaelpj — APPROVED** [review]

**Pull request merged**